File: generated_code/cpp/solution_1022_1.cpp
#include <vector>

using namespace std;

class Solution {
private:
    int totalPaths = 0;
    int emptyCells = 0;
    int startRow = 0, startCol = 0;
    int numRows, numCols;

    // Backtracking function to explore all valid paths
    void backtrack(vector<vector<int>>& grid, int r, int c, int visitedCount) {
        // Base case: Reached the ending square
        if (grid[r][c] == 2) {
            // Check if all non-obstacle cells (including the start cell) have been visited
            if (visitedCount == emptyCells + 1) {
                totalPaths++;
            }
            return;
        }

        // Mark the current cell as visited using an obstacle placeholder (-1)
        int temp = grid[r][c];
        grid[r][c] = -1;

        // 4-directional displacements
        int dRow[] = {-1, 1, 0, 0};
        int dCol[] = {0, 0, -1, 1};

        for (int i = 0; i < 4; ++i) {
            int nextR = r + dRow[i];
            int nextC = c + dCol[i];

            // Validate boundaries and check if the cell is unvisited/empty or the destination
            if (nextR >= 0 && nextR < numRows && nextC >= 0 && nextC < numCols) {
                if (grid[nextR][nextC] == 0 || grid[nextR][nextC] == 2) {
                    backtrack(grid, nextR, nextC, visitedCount + 1);
                }
            }
        }

        // Restore the cell state for other paths (backtrack step)
        grid[r][c] = temp;
    }

public:
    int uniquePathsIII(vector<vector<int>>& grid) {
        numRows = grid.size();
        numCols = grid[0].size();
        totalPaths = 0;
        emptyCells = 0;

        // Identify starting point and count total empty squares to visit
        for (int i = 0; i < numRows; ++i) {
            for (int j = 0; j < numCols; ++j) {
                if (grid[i][j] == 1) {
                    startRow = i;
                    startCol = j;
                } else if (grid[i][j] == 0) {
                    emptyCells++;
                }
            }
        }

        // Kick off backtracking starting at count 1 (counting the start cell itself)
        backtrack(grid, startRow, startCol, 1);

        return totalPaths;
    }
};

File: generated_code/python/solution_1022_1.py
from typing import List

class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        # Time Complexity: O(4^(m*n)) in the worst case, but heavily pruned. 
        # Given constraints (m * n <= 20), back tracking is highly efficient and optimal.
        # Space Complexity: O(m * n) for the recursion call stack.
        
        m, n = len(grid), len(grid[0])
        start_r, start_c = 0, 0
        empty_squares = 0
        
        # Step 1: Scan the grid to find the starting position and count total walkable squares
        for r in range(m):
            for c in range(n):
                if grid[r][c] == 1:
                    start_r, start_c = r, c
                    empty_squares += 1
                elif grid[r][c] == 0:
                    empty_squares += 1
                elif grid[r][c] == 2:
                    empty_squares += 1  # Include the ending square in the count
                    
        self.paths_count = 0
        
        # Step 2: Define backtracking function to explore paths
        def backtrack(r: int, c: int, remain: int):
            # Base Case: If we reach the target square
            if grid[r][c] == 2:
                # If all non-obstacle squares have been visited, we found a valid path
                if remain == 1:
                    self.paths_count += 1
                return
            
            # Mark the current square as visited by temporarily setting it as an obstacle
            temp = grid[r][c]
            grid[r][c] = -1
            
            # Explore 4-directional neighbors
            for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                nr, nc = r + dr, c + dc
                if 0 <= nr < m and 0 <= nc < n and grid[nr][nc] != -1:
                    backtrack(nr, nc, remain - 1)
                    
            # Backtrack: Restore the square's original value
            grid[r][c] = temp
            
        # Kickstart backtracking from the starting position
        backtrack(start_r, start_c, empty_squares)
        
        return self.paths_count